### PR TITLE
 Handle ingress domain updates to default cluster ingress

### PR DESCRIPTION
**Pull request by pravisankar**
- Delete duplicate finalizers section in default cluster ingress cr.
- Current code fetches needed info from installer configmap and tries to create default cluster ingress cr and ignores in case of already exists error. It won't handle the case where default cluster ingress cr exists but the ingress domain got changed. This pr handles this case.

### CODE REVIEW

**Comment by pravisankar** [comment]
@openshift/sig-network-edge PTAL

File: pkg/stub/handler.go [hunk 1]
```diff
@@ -62,16 +62,48 @@ func (h *Handler) EnsureDefaultClusterIngress() error {
...
+		},
+	}
+	err := sdk.Get(oldci)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return false, false, fmt.Errorf("failed to fetch existing default cluster ingress %s/%s, %v", ci.Namespace, ci.Name, err)
+		}
+		return false, true, nil
+	}
+	if oldci.Spec.IngressDomain != ci.Spec.IngressDomain {
```
Comment (ironcladlou) [inline pkg/stub/handler.go]: I think you need to deference the pointers to do what you mean here: https://play.golang.org/p/N1-AT4yXgH-

Comment (pravisankar) [reply pkg/stub/handler.go]: yes, thanks

**Comment by pravisankar** [comment]
@ironcladlou updated, ptal

**Comment by ironcladlou** [comment]
How was it tested?

**Comment by pravisankar** [comment]
Tested both create and update cases on 4.0 cluster
/hold cancel

**Comment by pravisankar** [comment]
@openshift/sig-network-edge updated, PTAL

File: pkg/stub/handler.go [hunk 2]
```diff
@@ -62,16 +62,49 @@ func (h *Handler) EnsureDefaultClusterIngress() error {
...
+		},
+	}
+	err := sdk.Get(oldci)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return false, nil, fmt.Errorf("failed to fetch existing default cluster ingress %s/%s, %v", ci.Namespace, ci.Name, err)
+		}
+		return false, nil, nil
+	}
+	if *oldci.Spec.IngressDomain != *ci.Spec.IngressDomain {
```
Comment (ironcladlou) [inline pkg/stub/handler.go]: If either of this is `nil` will the function panic?

Comment (pravisankar) [reply pkg/stub/handler.go]: Let me fix this

**Comment by ironcladlou** [comment]
Would like to test this once https://github.com/openshift/cluster-ingress-operator/pull/51 lands

File: pkg/stub/handler.go [hunk 3]
```diff
@@ -69,16 +69,56 @@ func (h *Handler) EnsureDefaultClusterIngress() error {
 		return err
 	}
 
-	err = sdk.Create(ci)
-	if err == nil || errors.IsAlreadyExists(err) {
-		if err == nil {
-			logrus.Infof("created default cluster ingress %s/%s", ci.Namespace, ci.Name)
+	changed, nci, err := checkClusterIngress(ci)
+	if err != nil {
```
Comment (imcsk8) [inline pkg/stub/handler.go]: I think this can use: clusterIngressStatusEqual
from PR: #38

Comment (Miciah) [reply pkg/stub/handler.go]: Maybe not `clusterIngressStatusEqual`, but the part that compares `.loadBalancer.Ingress` could be factored into a function that could be used both by `checkClusterIngress` and by `clusterIngressStatusEqual`.  Now that #47 is merged, I'll rebase #38 Tuesday and factor out a `LoadBalancerStatusEqual` function.  No need for that to block this PR though; I can change `checkClusterIngress` to use `LoadBalancerStatusEqual` if #48 merges before #38 is ready.

Comment (Miciah) [reply pkg/stub/handler.go]: Err, actually, `checkClusterIngress` is comparing only `.spec.ingressDomain` (a string) whereas `clusterIngressStatusEqual` is comparing `.status.loadBalancer` (a slice of a struct that has two string values), so there is really no opportunity for code re-use here as far as I see.

Comment (imcsk8) [reply pkg/stub/handler.go]: I might be confused but if the LoadBalancerIngress struct are different as being checked in your PR [1]  it means there's an update it should be enough for this purpouse.

[1] https://github.com/openshift/cluster-ingress-operator/pull/38/files#diff-9ae69f8dcadd8df44e0b10265f3a91a8R107

**Comment by ironcladlou** [comment]
Superseded by https://github.com/openshift/cluster-ingress-operator/pull/48

**Comment by Miciah** [comment]
> Superseded by #48

For the record, I think that was supposed to be #89.

**Pull request closed without merging**